3406 Develop ES Sweep Indexer

albertisfu opened the pull request:
This PR includes the implementation of the Sweep indexer as outlined in #3406.

It serially processes each supported model, indexing their instances. The supported models and the order of indexing are as follows:

```
[
    "audio.Audio",
    "people_db.Person",
    "search.OpinionCluster",
    "search.Opinion",
    "search.Docket",
    "search.RECAPDocument",
]
```
`ParentheticalGroup` is not yet supported, as the mapping for this search type may change following the improvements we need to implement there.

- For Audio, the command will index only **`Audio`** instances where **`processing_complete`** is **`True`**.
- For Person, the command will index **`Person`** instances (including all their Positions) where the property **`is_judge`** is **`True`**.
- For **`OpinionCluster`**, it will index all **`OpinionCluster`** instances from the database in bulk, without considering the Opinions at this stage to speed up the indexing process.
- For **`Opinion`**, upon reaching this stage, it will index all **`Opinion`** instances from the database in bulk. If new Opinions that belong to new OpinionClusters are added to the database, the indexing task will ensure the missing `OpinionClusters` are indexed in bulk before their `Opinions`.
- For **`Docket`**, it will index all **`Docket`** instances from the database in bulk, which also speed up the indexing as it does not require querying their `RECAPDocuments` at this stage.
- For **`RECAPDocument`**, it will index all **`RECAPDocument`** instances from the database in bulk. Similar to Opinions, if a new parent `Docket` is not yet indexed, it will be indexed before its `RECAPDocuments` to prevent orphaned documents in the index.

This new approach should be faster and more efficient, with the trade-off being that when indexing missing parent documents, they may be empty until the command reaches the child model as well.

The indexer has two modes:

- **`all`**: It will re-index all documents, even those already indexed. This is useful for correcting previously indexed documents, as we need to do now according to #3629.
- **`missing`**: After completing the first sweep indexer cycle, this mode can be enabled to process only documents that are not yet indexed. It works by checking the document's existence in its Elasticsearch index before scheduling it for indexing. I considered adding an email notification to inform us when the first cycle is finished so we can enable this option. The indexer currently only logs when it finishes: **`Sweep indexer 'all' cycle finished successfully.`** If there is a way to detect when this log is printed, that will signal us to stop it, enable the **`missing`** option, and restart the command; otherwise, it will start a new indexing cycle.
- This mode can be set in the following setting: **`ELASTICSEARCH_SWEEP_INDEXER_ACTION`**.

- Other available settings include:
    - **`ELASTICSEARCH_SWEEP_INDEXER_QUEUE`**: the Celery queue defined to run the sweep indexer tasks. Defaults to **`celery`**. This is useful for isolating these tasks from those triggered by signals, ensuring the indexer's throttling is not affected by signal tasks.
    - **`ELASTICSEARCH_SWEEP_INDEXER_CHUNK_SIZE`**: the chunk size for indexing documents in bulk into Elasticsearch. This number is also used for the **`min_items`** parameter in **`CeleryThrottle`**. Defaults to `50`.
    - **`ELASTICSEARCH_SWEEP_INDEXER_POLL_INTERVAL`**: used for **`poll_interval`** in **`CeleryThrottle`**, allowing us to adjust the indexing rate as needed.
- The sweep indexer can automatically resume where it left off, storing the last indexing status in Redis. This status is logged every 1000 processed items. So, if the indexer fails for any reason, upon restart, it will continue from the last model and document ID logged. It also records the cumulative number of documents indexed for each model. At the end of a whole cycle, it logs the total number of documents indexed. If **`missing`** was enabled, it logs only the number of documents that were missing and indexed.
- I have added the command to **`docker-entrypoint.sh`** to function as a daemon in a pod. I assume that if the indexer crashes, the pod will automatically be recreated, and the indexer will start again. Is that correct, or do I need to do something else to ensure the indexer can restart its execution after a failure?

Now, I'll confirm what we're missing in #3629 or in other indices, so this can be deployed and start a new full re-index.

Let me know what do you think.

mlissner commented on cl/search/management/commands/sweep_indexer.py:
```diff
@@ -0,0 +1,392 @@
+            models_stack = list_supported_models[::-1]
+            self.process_documents(models_stack)
+
+    def process_documents(
```
Maybe rename this as `process_model_stack`? It's a bit confusing to have `process_documents`, which calls `process_queryset`?

albertisfu replied:
sure this name is better, changed!

mlissner commented on cl/search/management/commands/sweep_indexer.py:
```diff
@@ -0,0 +1,392 @@
+        :return: None
+        """
+        testing_mode = self.options.get("testing_mode", False)
+        model_options = {
```
Could you combine this with the match statement above, so you just have one match statement or dictionary with all the data needed for each object?

albertisfu replied:
Yeah, I've moved the params within the match case statement in the previous method

mlissner commented on cl/settings/third_party/elasticsearch.py:
```diff
@@ -231,3 +231,20 @@
+ELASTICSEARCH_SWEEP_INDEXER_ACTION = env(
+    "ELASTICSEARCH_SWEEP_INDEXER_ACTION", default="all"
+)
+ELASTICSEARCH_SWEEP_INDEXER_QUEUE = env(
+    "ELASTICSEARCH_SWEEP_INDEXER_QUEUE", default="celery"
+)
```
I think we should make a special task queue for this. I'll deploy one called `es_sweep`. Can you set that as the default?

albertisfu replied:
Done, I've set `es_sweep` as the default queue!

mlissner reviewed (changes requested):
This feels generally good, and I think you've got the design where it should be. Just a few comments from me, and off to Eduardo for review.

albertisfu commented on cl/search/management/commands/sweep_indexer.py:
```diff
@@ -0,0 +1,392 @@
+    log_key = compose_indexer_redis_key()
+    stored_values = r.hgetall(log_key)
+
+    # Retrieve the number of documents processed dinamically.
```
Done!

albertisfu commented on cl/search/management/commands/sweep_indexer.py:
```diff
@@ -0,0 +1,392 @@
+    """
+
+    r = get_redis_interface("CACHE")
+    pipe = r.pipeline()
```
Yes, that's correct; it executes only one operation at a time. Removed.

albertisfu commented on cl/search/management/commands/sweep_indexer.py:
```diff
@@ -0,0 +1,392 @@
+)
+from cl.search.types import ESDocumentClassType
+
+list_supported_models = [
```
Sure! I've renamed the variable.

albertisfu commented on cl/search/management/commands/sweep_indexer.py:
```diff
@@ -0,0 +1,392 @@
+        documents_dict[document] = current_total
+
+    data_to_log = {
+        "document_name": document_name,
```
yeah a better name, changed!

albertisfu commented on cl/search/management/commands/sweep_indexer.py:
```diff
@@ -0,0 +1,392 @@
+    """Get the last document_name and ID indexed from Redis.
+    :return: The last document name and ID indexed.
+    """
+    r = get_redis_interface("CACHE")
```
Yeah, done!

albertisfu commented:
Thanks for your comments and suggestions. I've applied them.

ERosendo commented on cl/search/tasks.py:
```diff
@@ -1112,6 +1117,18 @@ def index_parent_or_child_docs(
                 child_instances = Opinion.objects.filter(pk__in=instance_ids)
+                # Get unique parent_ids for Opinions
+                parent_ids = list(
+                    set(
+                        Opinion.objects.filter(
+                            pk__in=instance_ids
+                        ).values_list("cluster_id", flat=True)
+                    )
```
is there a reason to use the set constructor instead of the [distinct method](https://docs.djangoproject.com/en/5.0/ref/models/querysets/#distinct) ?

albertisfu replied:
Yeah, I changed updated this too!

ERosendo commented on cl/search/tasks.py:
```diff
@@ -1098,9 +1098,14 @@ def index_parent_or_child_docs(
-                ).values_list("docket_entry__docket_id", flat=True)
+                # Get unique parent_ids for RECAPDocuments
+                parent_ids = list(
+                    set(
+                        RECAPDocument.objects.filter(
+                            pk__in=instance_ids
+                        ).values_list("docket_entry__docket_id", flat=True)
+                    )
```
same question about the distinct method here🤔.

albertisfu replied:
Oh yeah, I used a `set` because the `distinct` method was not working here due to the nested relation `docket_entry__docket_id`. But you're right is it possible to use `distinct()` here just using `order_by` and setting the field name. Changed.

ERosendo commented on cl/search/management/commands/sweep_indexer.py:
```diff
@@ -0,0 +1,418 @@
+        "date_time": datetime.now().isoformat(),
+    }
+    data_to_log.update(documents_dict)
+    log_info = cast(Mapping[str | bytes, int | str], data_to_log)
```
The `data_to_log` variable seems to use only `string` keys. Are keys of type `bytes` possible in this scenario? Or is the mapping not updated?

albertisfu replied:
Yeah, the mapping was not updated. I just fixed it. Thanks for noticing it.

ERosendo commented on cl/search/management/commands/sweep_indexer.py:
```diff
@@ -0,0 +1,418 @@
+
+
+def log_indexer_last_status(
+    model_name: str, document_pk: int, chunk_size: int, log_key: str
```
Instead of passing `log_key` as a parameter to this method, could we achieve the same result by using the `compose_indexer_redis_key` helper directly within the method to compute the `key`?

albertisfu replied:
Yeah, this is possible. Changed!

ERosendo commented on cl/search/management/commands/sweep_indexer.py:
```diff
@@ -0,0 +1,418 @@
+    return documents_processed
+
+
+def find_starting_model(models: list[str], target_model: str) -> int | None:
```
Considering the `supported_models` variable is global, can we simplify this function by accessing it directly instead of passing the list of models as a parameter?

albertisfu replied:
Yeah. I removed the models from the arguments and just retrieve them inside the method.

ERosendo commented on cl/search/management/commands/sweep_indexer.py:
```diff
@@ -0,0 +1,418 @@
+    :return: The index of the model if found, otherwise None.
+    """
+
+    for index, model_name in enumerate(models):
+        if model_name == target_model:
+            return index
+    return None
```
We can use the `index` method here: 

```suggestion
    try:
        return models.index(target_model)
    except ValueError:
        return None
```

albertisfu replied:
Thanks for the suggestion. I've applied it.

ERosendo commented:
The code looks good. We can merge the PR once we've addressed the latest comments.

albertisfu commented:
@ERosendo Thanks for your suggestions, I applied the changes.

@mlissner I've introduced a filter for the Dockets query to fix #3702, so now only Dockets that belong to `RECAP_SOURCES` will be indexed. I also updated `RECAP_SOURCES`, as I think it was outdated after the addition of the Harvard source.

I'll open an issue to apply this filter to indexing signals too, and another to remind us to build the script for removing content that is not part of RECAP from ES.
